### PR TITLE
W6 - D1 - thanh hoang - Unit Test 

### PR DESCRIPTION
**Pull request by at-thanhhoang**
Task #62

### CODE REVIEW

File: app/src/testAt-thanhhoang/java/vn/asiantech/internship/TestPasswordValidate.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import static org.mockito.Mockito.doReturn;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TestPasswordValidate {
+    @Spy
+    private Account mPasswordAccount;
+
+    @Test
+    public void testCasePasswordDifferentUsername() {
+        doReturn("gfD1bdDs").when(mPasswordAccount).getUsername();
```
Comment (at-dev-support-01) [inline app/src/testAt-thanhhoang/java/vn/asiantech/internship/TestPasswordValidate.java]: WARNING: [AvoidDuplicateLiterals] The String literal "gfD1bdDs" appears 7 times in this file; the first occurrence is on line 22
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

**Comment by at-thanhhoang** [comment]
@at-trungnguyen . em đã làm xong bài tập. anh review giúp em với. cảm ơn anh.

File: app/build.gradle [hunk 2]
```diff
@@ -69,5 +69,8 @@ dependencies {
     implementation 'com.android.support:cardview-v7:27.1.1'
     implementation 'com.github.bumptech.glide:glide:4.7.1'
     annotationProcessor 'com.github.bumptech.glide:compiler:4.7.1'
+
+    testImplementation 'org.mockito:mockito-core:2.7.22'
+    androidTestImplementation 'org.mockito:mockito-android:2.7.22'
```
Comment (at-trungnguyen) [inline app/build.gradle]: remove, no need

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/Account.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package asiantech.internship.summer.unit.test;
+
+public class Account {
+    private String mUsername;
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/Account.java]: no `m` in model class

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/Account.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    private String mPassword;
+
+    Account(){}
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/Account.java]: format code

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UnitTestActivity.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private EditText mEdtPassword;
+    private TextView mTvError;
+    private Button mBtnLogin;
+
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_unit_test);
+
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UnitTestActivity.java]: delete this line `23`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UnitTestActivity.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    private void initView() {
+        mEdtUsername = findViewById(R.id.edtUsername);
+        mEdtPassword = findViewById(R.id.edtPasswordUnitTest);
+        mTvError = findViewById(R.id.tvError);
+        mBtnLogin = findViewById(R.id.btnLogIn);
+    }
+
+    private void addListener() {
+        mBtnLogin.setOnClickListener(v -> {
+            Account account = new Account(mEdtUsername.getText().toString(), mEdtPassword.getText().toString());
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UnitTestActivity.java]: use `.trim()` 
`mEdtUsername.getText().toString().trim()`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
+package asiantech.internship.summer.unit.test;
+
+import java.util.ArrayList;
+import java.util.regex.Pattern;
+
+import asiantech.internship.summer.R;
+
+public class UtilValidate {
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: `public final class`

Comment (at-trungnguyen) [reply app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: define `private` constructor

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    private static Pattern mRegex = Pattern.compile("[$&+,:;=\\\\?@#|/'<>.^*()%!-]");
+
+    private static boolean isUsernameLength(String username) {
+        return username.length() > 6 && username.length() < 22;
+    }
+
+    private static boolean isUsernameLeastTwoUppercaseCharacterNotContinue(String username, int numberUppercase) {
+        ArrayList<Integer> arrPosition = new ArrayList<>();
+        for (int i = 0; i < username.length(); i++) {
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: don't count `username.length()` inside `for`, define variable.

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+public class UtilValidate {
+    private static Pattern mRegex = Pattern.compile("[$&+,:;=\\\\?@#|/'<>.^*()%!-]");
+
+    private static boolean isUsernameLength(String username) {
+        return username.length() > 6 && username.length() < 22;
+    }
+
+    private static boolean isUsernameLeastTwoUppercaseCharacterNotContinue(String username, int numberUppercase) {
+        ArrayList<Integer> arrPosition = new ArrayList<>();
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: `List<Integer> arrPosition = new ArrayList<>();`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        for (int i = 0; i < username.length(); i++) {
+            if (Character.isUpperCase(username.charAt(i))) {
+                arrPosition.add(i);
+            }
+        }
+        if (arrPosition.size() >= numberUppercase) {
+            for (int i = 0; i < arrPosition.size() - 1; i++) {
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: same above

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    private static boolean isUsernameNotSpecialAndSpaceCharacter(String username) {
+        return !username.contains(" ") && !mRegex.matcher(username).find();
+    }
+
+    private static boolean isUsernameMostTwoDigitContinue(String username) {
+        int count = 0;
+        ArrayList<Integer> arrPosition = new ArrayList<>();
+        for (int i = 0; i < username.length(); i++) {
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: same above

File: app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        return !Character.isUpperCase(c) && !Character.isDigit(c);
+    }
+
+    private static boolean isPasswordDifferentUsername(String password, String username) {
+        return !password.equals(username);
+    }
+
+    private static boolean isPasswordLeastTwoSpecialCharacter(String password) {
+        int count = 0;
+        for (int i = 0; i < password.length(); i++) {
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/unit/test/UtilValidate.java]: same above

File: app/src/at-thanhhoang/res/drawable/bg_view_unit_test.xml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <stroke
+        android:width="2dp"
+        android:color="@color/colorGrayDark" />
+    <corners android:radius="4dp" />
+    <solid android:color="@color/colorWhite" />
+</shape>
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/drawable/bg_view_unit_test.xml]: format and enter 1 line

File: app/src/at-thanhhoang/res/layout/activity_menu.xml [hunk 14]
```diff
@@ -56,5 +56,17 @@
             android:layout_marginLeft="20dp"
             android:textAllCaps="false"
             android:textSize="20sp"/>
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/layout/activity_menu.xml]: text size is `dp`

File: app/src/at-thanhhoang/res/layout/activity_menu.xml [hunk 15]
```diff
@@ -56,5 +56,17 @@
...
+            android:layout_marginTop="10dp"
+            android:id="@+id/btnUnitTest"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:text="@string/label_unit_test"
+            android:background="@color/colorBackgroundButtonMenuActivity"
+            android:layout_marginRight="20dp"
+            android:layout_marginLeft="20dp"
+            android:textAllCaps="false"
+            android:textSize="20sp"/>
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/layout/activity_menu.xml]: `dp` and put in `dimens.xml`

File: app/src/at-thanhhoang/res/layout/activity_unit_test.xml [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="30dp"
+        android:layout_marginRight="30dp"
+        android:layout_marginTop="20dp"
+        android:background="@drawable/bg_view_unit_test"
+        android:gravity="center"
+        android:hint="@string/label_error"
+        android:padding="15dp"
+        android:singleLine="false"
+        android:textSize="18dp" />
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/layout/activity_unit_test.xml]: define in `dimens.xml`

File: app/src/at-thanhhoang/res/layout/activity_unit_test.xml [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center_horizontal"
+        android:layout_marginTop="20dp"
+        android:background="@color/colorGreenLight"
+        android:paddingLeft="10dp"
+        android:paddingRight="10dp"
+        android:text="@string/label_login"
+        android:textAllCaps="false"
+        android:textSize="18dp" />
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/layout/activity_unit_test.xml]: define it

File: app/src/at-thanhhoang/res/layout/activity_unit_test.xml [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center_horizontal"
+        android:layout_marginTop="20dp"
+        android:text="@string/or_login_in_using_a_social_account"
+        android:textSize="15dp"
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/layout/activity_unit_test.xml]: define it

File: app/src/at-thanhhoang/res/values/colors.xml [hunk 19]
```diff
@@ -22,4 +22,9 @@
     <color name="colorYellow">#ffee58</color>
     <color name="colorBrown">#3e2723</color>
     <color name="colorGray">#f1f1f1</color>
+    <color name="colorGrayDark">#bdbdbd</color>
+    <color name="colorGreenLight">#8bc34a</color>
+    <color name="colorGreenDark">#66bb6a</color>
+    <color name="colorRed">#e53935</color>
+
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/values/colors.xml]: why space here? delete line 29

File: app/src/at-thanhhoang/res/values/strings.xml [hunk 20]
```diff
@@ -28,9 +28,29 @@
...
+    <string name="check_username_not_space_and_special_character">Username does not contain special characters and spaces</string>
+    <string name="check_username_most_two_digit_continue">Username have at most 2 digit continue</string>
+    <string name="check_username_not_start_uppercase_or_digit">Username do not start with uppercase or numbers</string>
+    <string name="check_password_different_username">Password and Username must be different</string>
+    <string name="check_password_least_two_special_or_digit">Password have at least two special characters or digit</string>
+    <string name="check_password_length_and_not_repeated_character_than_two_continue">Password have minimum length is 8 characters and can not be repeated 1 character more than 2 times in a row</string>
+    <string name="check_password_not_end_digit_or_special_character">Password do not end with digit or special characters</string>
+    <string name="check_password_least_three_uppercase_character_not_continue">Password have at least 3 uppercase characters but not consecutive</string>
+    <string name="check_login_success">Login success!!!</string>
+
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/values/strings.xml]: delete this line

File: app/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -1,5 +1,3 @@
 <resources>
     <string name="app_name">Intern18Summer</string>
-    <string name="label_sign_up_login">SignUp?</string>
-    <string name="label_login">Login</string>
-</resources>
+</resources>
```
Comment (at-trungnguyen) [inline app/src/main/res/values/strings.xml]: miss enter

File: app/src/testAt-thanhhoang/java/vn/asiantech/internship/TestPasswordValidate.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    @Spy
+    private Account mPasswordAccount;
+
+    @Test
+    public void testCasePasswordDifferentUsername() {
+        doReturn(USERNAME).when(mPasswordAccount).getUsername();
+        doReturn(USERNAME).when(mPasswordAccount).getPassword();
+        assertEquals(UtilValidate.isLogin(mPasswordAccount), R.string.check_password_different_username);
+    }
+
```
Comment (at-trungnguyen) [inline app/src/testAt-thanhhoang/java/vn/asiantech/internship/TestPasswordValidate.java]: delete this line 27

File: app/src/at-thanhhoang/res/values/dimens.xml [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+    <dimen name="menu_text_size">20dp</dimen>
+    <dimen name="unit_text_padding_edit_text">15dp</dimen>
+    <dimen name="unit_text_padding_button_login">10dp</dimen>
+    <dimen name="unit_text_text_size_large">18dp</dimen>
+    <dimen name="unit_text_text_size_small">15dp</dimen>
+    <dimen name="unit_text_margin">30dp</dimen>
+    <dimen name="unit_text_margin_top">20dp</dimen>
+    <dimen name="unit_text_height_image">100dp</dimen>
+    <dimen name="unit_text_weight_image">40dp</dimen>
+</resources>
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/values/dimens.xml]: enter 1 line

File: app/src/at-thanhhoang/res/layout/activity_unit_test.xml [hunk 24]
```diff
@@ -9,94 +9,94 @@
...
-            android:layout_height="40dp"
+            android:layout_width="@dimen/unit_text_height_image"
+            android:layout_height="@dimen/unit_text_weight_image"
             android:background="@drawable/img_google" />
 
         <ImageView
-            android:layout_width="100dp"
-            android:layout_height="40dp"
-            android:layout_marginLeft="10dp"
+            android:layout_width="@dimen/unit_text_height_image"
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/layout/activity_unit_test.xml]: this is optional, you can set directly `android:layout_marginLeft="10dp"`

File: app/src/at-thanhhoang/res/drawable/bg_view_unit_test.xml [hunk 25]
```diff
@@ -0,0 +1,9 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <stroke
+        android:width="2dp"
+        android:color="@color/colorGrayDark" />
+
+    <corners android:radius="4dp" />
+
+    <solid android:color="@color/colorWhite" />
+</shape>
```
Comment (at-trungnguyen) [inline app/src/at-thanhhoang/res/drawable/bg_view_unit_test.xml]: enter

File: app/src/main/res/values/strings.xml [hunk 26]
```diff
@@ -1,7 +1,3 @@
 <resources>
     <string name="app_name">Intern18Summer</string>
-    <string name="label_sign_up_login">SignUp?</string>
-    <string name="label_login">Login</string>
-    <string name="load_image">Load Image</string>
-    <string name="upload_image">Upload Image</string>
-</resources>
+</resources>
```
Comment (at-trungnguyen) [inline app/src/main/res/values/strings.xml]: enter

File: app/src/testAt-thanhhoang/java/vn/asiantech/internship/TestPasswordValidate.java (hunk 22 above)
Comment (at-trungnguyen) [reply app/src/testAt-thanhhoang/java/vn/asiantech/internship/TestPasswordValidate.java]: check again

**Comment by at-trungnguyen** [comment]
@at-thanhhoang Please check again.

**Pull request merged**